### PR TITLE
WIP: MJ + Angus private cluster combined PR for testing

### PR DESCRIPTION
**Pull request by asalkeld**
```release-note
NONE
```
/cc @mjudeikis @jim-minter 

this appears to be getting closer now:
- sync pod is getting updated
- healthz is working
- masters are coming up
- infra is not coming up yet ...
```                 
DEBU[2019-10-17T16:23:19+10:00]pkg/fakerp/fakerp.go:170 pkg/fakerp.GetDeployer.func1() PE IP Address 172.30.1.6                     
DEBU[2019-10-17T16:23:19+10:00]pkg/cluster/kubeclient/roundtrip.go:54 pkg/cluster/kubeclient.NewRestConfig() override kubeClient roundtripper with PrivateEndpoint dialer 
INFO[2019-10-17T16:23:20+10:00]pkg/cluster/healthcheck.go:56 pkg/cluster.(*Upgrade).WaitForHealthzStatusOk() waiting for API server healthz               
DEBU[2019-10-17T16:23:20+10:00]pkg/fakerp/proxy.go:39 pkg/fakerp.(*Server).configureProxyDialer.func1() dial 172.30.1.6:443                          
INFO[2019-10-17T16:23:24+10:00]pkg/cluster/update_syncpod.go:9 pkg/cluster.(*Upgrade).CreateOrUpdateSyncPod() updating sync pod                            
DEBU[2019-10-17T16:23:24+10:00]pkg/fakerp/proxy.go:39 pkg/fakerp.(*Server).configureProxyDialer.func1() dial 172.30.1.6:443                          
DEBU[2019-10-17T16:23:24+10:00]pkg/fakerp/proxy.go:39 pkg/fakerp.(*Server).configureProxyDialer.func1() dial 172.30.1.6:443                          
DEBU[2019-10-17T16:23:25+10:00]pkg/fakerp/proxy.go:39 pkg/fakerp.(*Server).configureProxyDialer.func1() dial 172.30.1.6:443                          
DEBU[2019-10-17T16:23:25+10:00]pkg/fakerp/proxy.go:39 pkg/fakerp.(*Server).configureProxyDialer.func1() dial 172.30.1.6:443                          
DEBU[2019-10-17T16:23:25+10:00]pkg/fakerp/proxy.go:39 pkg/fakerp.(*Server).configureProxyDialer.func1() dial 172.30.1.6:443                          
INFO[2019-10-17T16:23:26+10:00]pkg/cluster/ready.go:19 pkg/cluster.(*Upgrade).WaitForNodesInAgentPoolProfile() waiting for master-000000 to be ready        
DEBU[2019-10-17T16:23:26+10:00]pkg/fakerp/proxy.go:39 pkg/fakerp.(*Server).configureProxyDialer.func1() dial 172.30.1.6:443                          

```

### CODE REVIEW

File: cmd/createorupdate/createorupdate.go [hunk 1]
```diff
@@ -305,13 +305,19 @@ func (c Client) Create(ctx context.Context) (string, error) {
...
-	// case "2019-10-27-preview" forward to default
-	default:
+	case "2019-10-27-preview":
 		oc, err := c.createOrUpdatev20191027preview(ctx)
 		if err != nil {
 			return "", err
 		}
 		consoleURL = *oc.Properties.PublicHostname
+	// case "2019-09-30-preview" forward to default
+	default:
```
Comment (mjudeikis) [inline cmd/createorupdate/createorupdate.go]: Follow-up: make privatecluster api default and introduce 2 testing modes.

File: cmd/test/test.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+		return fmt.Errorf("error configuring proxy")
+	}
+
+	roundtrippers.DialHook = func(network, address string) (net.Conn, error) {
+		s.log.Debugf("dial %s", address)
+		/* #nosec - connecting to external IP of a FakeRP cluster, expect self signed cert */
+		c, err := tls.Dial("tcp", s.testConfig.ProxyURL, &tls.Config{
+			RootCAs:      roots,
+			Certificates: []tls.Certificate{s.testConfig.ProxyCertificate},
+			// TOFIX: Current certificate does not contain
```
Comment (mjudeikis) [inline cmd/test/test.go]: Follow-up

File: cmd/test/test.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
+package main
```
Comment (mjudeikis) [inline cmd/test/test.go]: The fact that you have this and I have my version of this, I ok to commit this to the code base as it makes development easier. Can we have some header comments what this script is intended to do? and rename `test` to something like `pe-test`?

File: hack/cert-setup.sh [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+# ./makecert.sh foo@foo.com
```
Comment (mjudeikis) [inline hack/cert-setup.sh]: Follow-up - move to go

File: pkg/cluster/kubeclient/roundtrip.go [hunk 5]
```diff
@@ -1,161 +1,88 @@
...
+	} else {
+		tlsConfig, err := restclient.TLSConfigFor(restconfig)
 		if err != nil {
-			rt.Log.Warnf("%#v: not retrying", err)
+			return nil, err
 		}
 
-		return
+		restconfig.WrapTransport = roundtrippers.NewPrivateEndpoint(log, cs, disableKeepAlives, testConfig, tlsConfig)
+		if testConfig.RunningUnderTest {
```
Comment (mjudeikis) [inline pkg/cluster/kubeclient/roundtrip.go]: This is partially my code base, but reading it with fresh eyes - we need much better comment here.

File: pkg/cluster/types.go [hunk 6]
```diff
@@ -139,6 +140,15 @@ func NewSimpleUpgrader(ctx context.Context, log *logrus.Entry, cs *api.OpenShift
 	return u, nil
 }
 
+func (u *Upgrade) ReloadKubeClient(disableKeepAlives bool) error {
```
Comment (mjudeikis) [inline pkg/cluster/types.go]: Comment why we need this (as @jim-minter asked in other PR)

File: pkg/fakerp/config.go [hunk 7]
```diff
@@ -22,12 +24,40 @@ const (
 )
 
 func GetTestConfig() api.TestConfig {
+	// load proxy configuration for tests
+	var cert tls.Certificate
+	var ca []byte
+
+	// TODO: improve this
```
Comment (mjudeikis) [inline pkg/fakerp/config.go]: follow-up

File: pkg/fakerp/proxy.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		s.log.Debugf("%s is %s", proxyEnvName, s.testConfig.ProxyURL)
+	}
+
+	roots := x509.NewCertPool()
+	if ok := roots.AppendCertsFromPEM(s.testConfig.ProxyCa); !ok {
+		return fmt.Errorf("error configuring proxy")
+	}
+
+	roundtrippers.DialHook = func(network, address string) (net.Conn, error) {
+		s.log.Debugf("dial %s", address)
```
Comment (mjudeikis) [inline pkg/fakerp/proxy.go]: remove this - too noisy

File: test/e2e/standard/sanitychecker.go [hunk 9]
```diff
@@ -63,7 +64,12 @@ func NewSanityChecker(ctx context.Context, log *logrus.Entry, cs *internalapi.Op
 	if err != nil {
 		return nil, err
 	}
-	scc.Client, err = openshift.NewClientSet(log, cs)
+
+	testConfig := fakerp.GetTestConfig()
+
+	// TODO: Set PE IP here to enable tests to use PE
```
Comment (mjudeikis) [inline test/e2e/standard/sanitychecker.go]: This should be failing now?

File: test/management/data/start.sh [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+export HOME=/root
+cd
+
+yum install git -y
+
+curl -s https://storage.googleapis.com/golang/go1.11.6.linux-amd64.tar.gz | tar -C /usr/local -xz
+PATH=$PATH:/usr/local/go/bin
+
+go get -u github.com/mjudeikis/openshift-azure/cmd/proxy
```
Comment (mjudeikis) [inline test/management/data/start.sh]: follow-up - move to master when code is merged

**Comment by mjudeikis** [comment]
follow-up:
1.make privatecluster api default and introduce 2 testing modes.
2. Fix certificate SAN names (https://github.com/openshift/openshift-azure/pull/2006/files#diff-52f960cb47c501a0672dffb658162183R63)
3. Cert-setup - move to go
4. func GetTestConfig() api.TestConfig - refactor proxy cert reading

File: pkg/util/roundtrippers/healthcheck.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+		DialTLS: func(network, addr string) (net.Conn, error) {
+			_, port, err := net.SplitHostPort(addr)
+			if err != nil {
+				return nil, err
+			}
+			// This does the same thing as roundtrippers.NewPrivateEndpoint
+			if privateEndpoint != nil {
+				tlsConfig.InsecureSkipVerify = true
+				if testConfig.RunningUnderTest {
+					tlsConfig.ServerName = "172.30.0.1" // TODO don't hardcode this..
```
Comment (mjudeikis) [inline pkg/util/roundtrippers/healthcheck.go]: why we can use `host` from the split here?

**Comment by asalkeld** [comment]
roundtripper merged, so closing this

**Pull request closed without merging**